File: src/21.py
import sys
import fileinput


PLAYER_1_START = 8
PLAYER_2_START = 7

# PLAYER_1_START = 4
# PLAYER_2_START = 8


def solve_1(p1_start, p2_start):
    dice = create_dice()
    p1_pos = p1_start
    p2_pos = p2_start
    p1_score = 0
    p2_score = 0
    nb_rolls = 0

    while True:
        s = 0
        for _ in range(3):
            nb_rolls += 1
            d = next(dice)
            s += d
        p1_pos = (p1_pos + s - 1)%10 + 1
        p1_score += p1_pos
        if p1_score >= 1000:
            break

        s = 0
        for _ in range(3):
            nb_rolls += 1
            d = next(dice)
            s += d
        p2_pos = (p2_pos + s - 1)%10 + 1
        p2_score += p2_pos
        if p2_score >= 1000:
            break

    return nb_rolls*p2_score


P1 = 0
P2 = 1
POSSIBLE_MOVES = [3, 4, 5, 6, 7, 8, 9,]
UNIVERSES_FOR_MOVES = {3: 1, 4: 3, 5: 6, 6: 7, 7: 6, 8: 3, 9: 1}


def solve_2(p1_start, p2_start):
    """

    `(0 0) -- [1 j1 (4 21) (8 21)]`
    state = (1, P1, (4, 21), (8, 21))
    """
    wins = [0, 0]
    to_visit = [(1, P1, (p1_start, 21), (p2_start, 21))]
    m = {}
    find_wins(wins, to_visit)
    return max(wins)


def find_wins(wins, to_visit):
    while to_visit:
        state = to_visit.pop()
        # print(wins, state, to_visit)

        next_states = generate_next(state)
        player = turn(state)
        all_wins = True
        for ns in next_states:
            if goal(player, ns) > 0:
                to_visit.append(ns)
                all_wins = False
            else:
                wins[player] += nb_universes(ns)

    return wins, to_visit


def generate_next(state):
    result = []
    player = turn(state)
    other = other_player(player)
    n = nb_universes(state)
    p = pos(player, state)
    op = pos(other, state)
    g = goal(player, state)
    og = goal(other, state)
    for pm in POSSIBLE_MOVES:
        new_n = n*UNIVERSES_FOR_MOVES[pm]
        new_pos = advance(p, pm)
        new_goal = g - new_pos
        if player == P1:
            result.append((new_n, P2, (new_pos, new_goal), (op, og)))
        else:
            result.append((new_n, P1, (op, og), (new_pos, new_goal)))
    return result


def other_player(player):
    if player == P1:
        return P2
    return P1

def advance(start, move):
    return (start + move - 1)%10 + 1

def nb_universes(state):
    return state[0]

def turn(state):
    return state[1]

def pos(player, state):
    return state[2+player][0]

def goal(player, state):
    return state[2+player][1]

def create_dice():
    n = 1
    while True:
        yield n
        n += 1
        if n > 100:
            n -= 100


if __name__ == '__main__' and not sys.flags.interactive:
    print(solve_1(PLAYER_1_START, PLAYER_2_START))
    print(solve_2(PLAYER_1_START, PLAYER_2_START))

File: src/21.lua
PLAYER_1_START = 8
PLAYER_2_START = 7
-- PLAYER_1_START = 4
-- PLAYER_2_START = 8

function solve_1(p1_start, p2_start)
  local roll_dice = create_dice()
  local positions = {p1_start, p2_start}
  local scores = {0, 0}
  local nb_rolls = 0

  function roll_3_times()
    local result = 0
    for i=1,3 do
      nb_rolls = nb_rolls + 1
      result = result + roll_dice()
    end
    return result
  end

  done = false
  while not done do
    for player=1,2 do
      local rolls = roll_3_times()
      positions[player] = (positions[player] + rolls - 1)%10 + 1
      scores[player] = scores[player] + positions[player]
      if scores[player] >= 1000 then
        done = true
        break
      end
    end
  end

  return nb_rolls*math.min(scores[1], scores[2])
end

function solve_2(p1_start, p2_start)
  local wins = find_wins({}, p1_start, 0, p2_start, 0)
  return math.max(wins[1], wins[2])
end

function find_wins(m, p1_pos, p1_score, p2_pos, p2_score)
  if p1_score >= 21 then
    return {1, 0}
  end
  if p2_score >= 21 then
    return {0, 1}
  end

  local key = string.format("%d %d %d %d", p1_pos, p1_score, p2_pos, p2_score)
  if m[key] then
    return m[key]
  end

  local result = {0, 0}
  for d1=1,3 do
    for d2=1,3 do
      for d3=1,3 do
        new_p1_pos = (p1_pos + d1 + d2 + d3 - 1)%10 + 1
        new_p1_score = p1_score + new_p1_pos

        new_wins = find_wins(m, p2_pos, p2_score, new_p1_pos, new_p1_score)
        result[1] = result[1] + new_wins[2]
        result[2] = result[2] + new_wins[1]
      end
    end
  end
  m[key] = result
  return result
end

function create_dice()
  local n = 0
  return function()
      n = n + 1
      if n > 100 then
        n = n - 100
      end
      return n
  end
end

print(solve_1(PLAYER_1_START, PLAYER_2_START))
print(solve_2(PLAYER_1_START, PLAYER_2_START))
